### PR TITLE
Java.Net.URLEncoder/URLDecoder and Java.Util.BitSet

### PR DESCRIPTION
**Pull request by truongchauhien**

### CODE REVIEW

File: java/io/IOException/IOExceptionTest.cpp [hunk 1]
```diff
@@ -52,10 +52,10 @@ TEST (JavaLang, IOExceptionConstructor) {
     ASSERT_STR("IOException with the specified message", ioExceptionWithCause.getCause()->getCause()->getMessage().toString());
 }
 
-TEST (JavaLang, IOExceptionTryCatch) {
+TEST (JavaIO, IOExceptionTryCatch) {
     try {
         throw IOException("Throw IOException");
     } catch (Exception e) {
```
Comment (dthongvl) [inline java/io/IOException/IOExceptionTest.cpp]: Please catch **IOException &**

File: java/io/UnsupportedEncodingException/UnsupportedEncodingExceptionTest.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Constructs a new UnsupportedEncodingException with the specified detail message.
+    UnsupportedEncodingException unsupportedEncodingExceptionWithMessage = UnsupportedEncodingException("UnsupportedEncodingException with the specified message");
+    ASSERT_STR("UnsupportedEncodingException with the specified message", unsupportedEncodingExceptionWithMessage.getMessage().toString());
+}
+
+TEST(JavaIO, UnsupportedEncodingExceptionTryCatch) {
+    try {
+        throw UnsupportedEncodingException("Throw UnsupportedEncodingException");
+    } catch (UnsupportedEncodingException ex) {
```
Comment (dthongvl) [inline java/io/UnsupportedEncodingException/UnsupportedEncodingExceptionTest.cpp]: You can improve with catch by reference

File: java/lang/Long/Long.cpp [hunk 3]
```diff
@@ -331,20 +331,16 @@ int Long::numberOfLeadingZeros(long i) {
...
-        return 64;
-    }
-
+    int x, y;
+    if (i == 0) return 64;
     int n = 63;
-    y = (int)i;
-    n = n -32;
-    x = y;
+    y = static_cast<int>(i); if (y != 0) { n = n -32; x = y; }
```
Comment (dthongvl) [inline java/lang/Long/Long.cpp]: Please help me fix this style

File: java/lang/Long/Long.cpp [hunk 4]
```diff
@@ -331,20 +331,16 @@ int Long::numberOfLeadingZeros(long i) {
  * @return int
  */
 int Long::numberOfTrailingZeros(long i) {
-    long x, y;
-    if (i == 0) {
-        return 64;
-    }
-
+    int x, y;
+    if (i == 0) return 64;
```
Comment (dthongvl) [inline java/lang/Long/Long.cpp]: Please use bracket

File: java/util/BitSet/BitSet.cpp [hunk 5]
```diff
@@ -0,0 +1,797 @@
...
+int BitSet::wordIndex(int bitIndex) {
+    // indexOfWord = bitIndex / (2 ^ addressBitsPerWord)
+    // (2 ^ addressBitsPerWord) is size of word.
+    int indexOfWord = bitIndex >> ADDRESS_BITS_PER_WORD;
+    return indexOfWord;
+}
+
+void BitSet::checkRange(int fromIndex, int toIndex) {
+    if (fromIndex < 0) {
+        throw IndexOutOfBoundsException(String("fromIndex < 0: ") +
```
Comment (dthongvl) [inline java/util/BitSet/BitSet.cpp]: Is there anyway to shorten this? I think it would be great if you do that

**Review by dthongvl — CHANGES_REQUESTED** [review]

File: java/util/BitSet/BitSet.cpp (hunk 5 above)
Comment (truongchauhien) [reply java/util/BitSet/BitSet.cpp]: Fixed, thanks!

File: java/io/UnsupportedEncodingException/UnsupportedEncodingExceptionTest.cpp (hunk 2 above)
Comment (truongchauhien) [reply java/io/UnsupportedEncodingException/UnsupportedEncodingExceptionTest.cpp]: OK, I will remember this tip, thanks!

File: java/io/IOException/IOExceptionTest.cpp (hunk 1 above)
Comment (truongchauhien) [reply java/io/IOException/IOExceptionTest.cpp]: Fixed, thanks!

File: java/util/BitSet/BitSet.cpp [hunk 6]
```diff
@@ -0,0 +1,797 @@
...
+void BitSet::expandTo(int wordIndex) {
+    int wordsRequired = wordIndex + 1;
+    if (this->wordsInUse < wordsRequired) {
+        this->ensureCapacity(wordsRequired);
+        this->wordsInUse = wordsRequired;
+    }
+}
+
+void BitSet::ensureCapacity(int wordsRequired) {
+    if (this->words.length < wordsRequired) {
```
Comment (dthongvl) [inline java/util/BitSet/BitSet.cpp]: Please use early return

File: java/net/URLEncoder/URLEncoderTest.cpp [hunk 7]
```diff
@@ -28,14 +28,17 @@ extern "C" {
...
 using namespace Java::Lang;
 
-TEST(JavaNet, UrlEncodeString) {
-//    unicode target = (unicode) "Quán ăn";
-//    String result = UrlEncoder::encode(target);
-//    unicode expect =  "Qu%c3%a1n+%c4%83n";
-//    ASSERT_STR(expect, result.toString());
-}
\ No newline at end of file
+TEST(JavaNet, URLEncodeString) {
```
Comment (dthongvl) [inline java/net/URLEncoder/URLEncoderTest.cpp]: I think you should divide into 2 test cases
EncodeString and EncodeStringWithEncoding or something

File: java/util/BitSet/BitSet.cpp [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+    if (indexOfWord >= this->wordsInUse) {
+        return;
+    }
+
+    this->words[indexOfWord] &= ~(1L << bitIndex);
+    this->recalculateWordsInUse();
+}
+
+void BitSet::clear(int fromIndex, int toIndex) {
+    checkRange(fromIndex, toIndex);
```
Comment (dthongvl) [inline java/util/BitSet/BitSet.cpp]: Please help me add **this->**

File: java/util/BitSet/BitSet.cpp [hunk 9]
```diff
@@ -0,0 +1,797 @@
...
+
+    if (fromIndex == toIndex) {
+        return;
+    }
+
+    int startWordIndex = wordIndex(fromIndex);
```
Comment (dthongvl) [inline java/util/BitSet/BitSet.cpp]: Please help me add **this->**

**Review by dthongvl — CHANGES_REQUESTED** [review]

File: java/util/BitSet/BitSet.cpp (hunk 9 above)
Comment (truongchauhien) [reply java/util/BitSet/BitSet.cpp]: Because `wordIndex` is a private static method, so should we use `this->`?

Comment (dthongvl) [reply java/util/BitSet/BitSet.cpp]: I think we use **this->** to show that it's not from outside class

Comment (truongchauhien) [reply java/util/BitSet/BitSet.cpp]: OK, thanks! Fixed!

File: java/util/BitSet/BitSet.cpp (hunk 6 above)
Comment (truongchauhien) [reply java/util/BitSet/BitSet.cpp]: Fixed! Thanks!

File: java/util/BitSet/BitSet.cpp (hunk 8 above)
Comment (truongchauhien) [reply java/util/BitSet/BitSet.cpp]: Fixed! Thanks!

File: java/net/URLEncoder/URLEncoderTest.cpp (hunk 7 above)
Comment (truongchauhien) [reply java/net/URLEncoder/URLEncoderTest.cpp]: OK, thanks!

File: java/lang/Long/Long.cpp (hunk 4 above)
Comment (truongchauhien) [reply java/lang/Long/Long.cpp]: Fixed.

File: java/lang/Long/Long.cpp (hunk 3 above)
Comment (truongchauhien) [reply java/lang/Long/Long.cpp]: Fixed.

**Review by dthongvl — APPROVED** [review]

**Pull request merged**